### DISCUSSION TITLE
TestClient crashes on websocket

### DISCUSSION THREAD

**Discussion by johnsxiongs** [fastapi-users/fastapi-users#436] [Q&A]
Here is my test code
```
import databases
import sqlalchemy
from fastapi import FastAPI, Depends
from fastapi import WebSocket, Request
from fastapi.testclient import TestClient
from fastapi_users import FastAPIUsers, models
from fastapi_users.authentication import JWTAuthentication
from fastapi_users.db import SQLAlchemyBaseUserTable, SQLAlchemyUserDatabase
from sqlalchemy.ext.declarative import DeclarativeMeta, declarative_base

DATABASE_URL = "sqlite:///./test.db"
SECRET = "SECRET"


class User(models.BaseUser):
    pass


class UserCreate(models.BaseUserCreate):
    pass


class UserUpdate(User, models.BaseUserUpdate):
    pass


class UserDB(User, models.BaseUserDB):
    pass


database = databases.Database(DATABASE_URL)
Base: DeclarativeMeta = declarative_base()


class UserTable(Base, SQLAlchemyBaseUserTable):
    pass


engine = sqlalchemy.create_engine(
    DATABASE_URL, connect_args={"check_same_thread": False}
)
Base.metadata.create_all(engine)

users = UserTable.__table__
user_db = SQLAlchemyUserDatabase(UserDB, database, users)


def on_after_register(user: UserDB, request: Request):
    print(f"User {user.id} has registered.")


def on_after_forgot_password(user: UserDB, token: str, request: Request):
    print(f"User {user.id} has forgot their password. Reset token: {token}")


jwt_authentication = JWTAuthentication(
    secret=SECRET, lifetime_seconds=3600, tokenUrl="/auth/jwt/login"
)

app = FastAPI()
fastapi_users = FastAPIUsers(
    user_db,
    [jwt_authentication],
    User,
    UserCreate,
    UserUpdate,
    UserDB,
)
app.include_router(
    fastapi_users.get_auth_router(jwt_authentication), prefix="/auth/jwt", tags=["auth"]
)
app.include_router(
    fastapi_users.get_register_router(on_after_register), prefix="/auth", tags=["auth"]
)
app.include_router(
    fastapi_users.get_reset_password_router(
        SECRET, after_forgot_password=on_after_forgot_password
    ),
    prefix="/auth",
    tags=["auth"],
)
app.include_router(fastapi_users.get_users_router(), prefix="/users", tags=["users"])


@app.websocket("/ws")
async def websocket_endpoint(self, websocket: WebSocket, user: User = Depends(fastapi_users.get_current_user)):
    print(user)
    await websocket.accept()
    try:
        while True:
            await asyncio.sleep(1)
            await websocket.send_json('hello123')
    except Exception as e:
        logger.info(e)


@app.on_event("startup")
async def startup():
    await database.connect()


@app.on_event("shutdown")
async def shutdown():
    await database.disconnect()


client = TestClient(app)

headers = {
    'Authorization': 'Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyX2lkIjoiZGQzNjc1NjUtZWEwOC00MTExLTk4YzQtN2I3NThiNmE1ZDYxIiwiYXVkIjoiZmFzdGFwaS11c2VyczphdXRoIiwiZXhwIjoxNjA5ODE0MDc1fQ.eTrTk9PswWkA9uEAcDkLp-u_dDe2UAN-gbNPcUyPTSM'
}


def test_websocket_handle_disconnection():
    with client.websocket_connect("/ws", headers=headers) as connection:
        while True:
            data = connection.receive_text()
            print(data)


test_websocket_handle_disconnection()

```

**Comment by johnsxiongs** [comment]
the error:

`TypeError: __call__() missing 1 required positional argument: 'request'`

**Comment by johnsxiongs** [comment]
@frankie567 
Is there an example about fastapi-users  use  in websockets?
Thanks!

**Comment by frankie567** [comment]
Websockets are not supported out-of-the box. However, I described a way to do it in #295:

> Actually, you could reuse the `CookieAuthentication` class logic. It has a `__call__` method which expects two parameters : credentials (`str`) and a database adapter:
> 
> https://github.com/frankie567/fastapi-users/blob/9e499791deae8d7c4c25f0ef6a786f493768fff0/fastapi_users/authentication/cookie.py#L61-L63
> 
> So you can do something like this, by reusing the components of `fastapi-users`:
> 
> ```python
> async def websocket_auth(websocket: WebSocket):
>     try:
>         cookie = websocket._cookies['fastapiusersauth']
>         user = await cookie_authentication(cookie, user_db)
>         # User is authenticated, you can also check if he is active
>         if user and user.is_active:
>             return user
> 
>         # The credentials are invalid, expired, or the user does not exist or inactive
>         return None
> ```
> 
> This way, you ensure the user does exist in database and `is_active`. Besides, you have the user object in your function, so you can do useful things like ensuring he is asking for his own data.

`websocket_auth` should be called at the beginning of your websocket handler. You can do roughly the same with `jwt_authentication`.

However, I've never tested it personally so it may need some tinkering there.

**Reply by johnsxiongs** [reply]
I'm sorry I don't  get understand about this. 
Would you mind finshing a full example code there!

**Comment by frankie567** [comment]
Ok, so, you wrote this:

```py
@app.websocket("/ws")
async def websocket_endpoint(self, websocket: WebSocket, user: User = Depends(fastapi_users.get_current_user)):
    print(user)
    await websocket.accept()
    try:
        while True:
            await asyncio.sleep(1)
            await websocket.send_json('hello123')
    except Exception as e:
        logger.info(e)
```

You can't use `fastapi_users.get_current_user` here because it expects to find a `Request` object, not a `WebSocket`. You should retrieve the token in the headers yourself and feed it to `jwt_authentication`.

```py
from fastapi import Header
from fastapi.security.utils import get_authorization_scheme_param


@app.websocket("/ws")
async def websocket_endpoint(self, websocket: WebSocket, authorization: str = Header()):
  scheme, param = get_authorization_scheme_param(authorization)  # "Bearer XXXXXX" -> "Bearer", "XXXXXX"
  user = await jwt_authentication(param, user_db)
  # We get a user: open the socket and do your things
  if user:
    await websocket.accept()
    try:
        while True:
            await asyncio.sleep(1)
            await websocket.send_json('hello123')
    except Exception as e:
        logger.info(e)
  # No user: close it
  else:
    await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
```

Actually, this is a limitation of `fastapi`, an issue was raised about this some days ago: https://github.com/tiangolo/fastapi/issues/2587 I'll follow this and try to help. If it gets solved, websockets should work out-of-the-box with `fastapi-users`.

**Reply by mrchoke** [reply]
> ```python
> from fastapi import Header
> from fastapi.security.utils import get_authorization_scheme_param
> 
> 
> @app.websocket("/ws")
> async def websocket_endpoint(self, websocket: WebSocket, authorization: str = Header()):
>   scheme, param = get_authorization_scheme_param(authorization)  # "Bearer XXXXXX" -> "Bearer", "XXXXXX"
>   user = await jwt_authentication(param, user_db)
>   # We get a user: open the socket and do your things
>   if user:
>     await websocket.accept()
>     try:
>         while True:
>             await asyncio.sleep(1)
>             await websocket.send_json('hello123')
>     except Exception as e:
>         logger.info(e)
>   # No user: close it
>   else:
>     await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
> ```

it not work with current   version ?

**Reply by peter65374** [reply]
current version we could not 'from fastapi-user.authentication import JWTAuthentication', so the above code has to be adapted to the new version to call for token verification.

**Comment by peter65374** [comment]
This is FastAPI websocket endpoint.
```python
@simple_frontend.websocket("/api/v1/protected_websocket")
async def protected_websocket_endpoint(websocket: WebSocket, user_manager: UserManager = Depends(get_user_manager)):
try:
        await websocket.accept()
        front_logger.debug(f"websocket /api/v1/protected_websocket handshake is established.")
        # authenticate JWT token for user verification
        auth_token = await websocket.receive_json()
        # auth_token format as below
        '''{
            "type": "authorization",
            "token": "{JWT token}"
        }'''
        token = auth_token.get('token')
        front_logger.info(f"receive auth token: {token}")
        if not token:
            # front_logger.error(f"User authentication failed.")
            raise WebSocketException(code=status.WS_1008_POLICY_VIOLATION, reason=f"User authentication failed.")
        user = await websocket_auth_user(token, user_manager)
        if not user:
            # front_logger.error(f"User authentication failed.")
            raise WebSocketException(code=status.WS_1008_POLICY_VIOLATION, reason=f"User authentication failed.")
        front_logger.info(f"User {user.id} authenticated.")

```
get_user_manager(...) and get_jwt_strategy() , you could directly get in detail from the documentation of fastapi-user full example at https://fastapi-users.github.io/fastapi-users/12.0/configuration/full-example/

The below is function websocket_auth_user to authenticate jwt or cookie token.
```python
async def websocket_auth_user(token: str, user_manager: UserManager) -> Optional[User]:
    try:
        # cookie = websocket._cookies['fastapiusersauth']
        jwt_strategy=get_jwt_strategy()
        user = await jwt_strategy.read_token(token=token, user_manager=user_manager)
        # The credentials are invalid, expired, or the user does not exist or inactive
        if not user:
            return None
        # User is authenticated, you can also check if he is active
        if user and user.is_active:
            return user
    except Exception as e:
        print(e)
        return None
```

**Reply by Tur8008** [reply]
Thank's for this workable solution!
